# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFocusManager.cpp
# Commit: 21cb609fe669
# Full Hash: 21cb609fe6690fbad6c6167e10ee2e7f9804a593
# Author: David Zbarsky <dzbarsky@gmail.com>
# Date: 2012-10-16 08:09:24
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Backout bug 779684 for topcrash
# ==============================================================================

diff -r a96deba7de41 -r 21cb609fe669 dom/base/nsFocusManager.cpp
--- a/dom/base/nsFocusManager.cpp	Mon Oct 15 01:42:40 2012 -0400
+++ b/dom/base/nsFocusManager.cpp	Mon Oct 15 14:35:50 2012 -0400
@@ -31,7 +31,7 @@
 #include "nsLayoutUtils.h"
 #include "nsIPresShell.h"
 #include "nsIContentViewer.h"
-#include "nsFrameIterator.h"
+#include "nsFrameTraversal.h"
 #include "nsObjectFrame.h"
 #include "nsEventDispatcher.h"
 #include "nsEventStateManager.h"
@@ -2270,8 +2270,15 @@
         if (nodeValue.Length() == (uint32_t)startOffset && !isFormControl &&
             startContent != aDocument->GetRootElement()) {
           // Yes, indeed we were at the end of the last node
-          nsFrameIterator frameTraversal(presContext, startFrame,
-                                         eLeaf, nsFrameIterator::FLAG_FOLLOW_OUT_OF_FLOW);
+          nsCOMPtr<nsIFrameEnumerator> frameTraversal;
+          nsresult rv = NS_NewFrameTraversal(getter_AddRefs(frameTraversal),
+                                             presContext, startFrame,
+                                             eLeaf,
+                                             false, // aVisual
+                                             false, // aLockInScrollView
+                                             true      // aFollowOOFs
+                                             );
+          NS_ENSURE_SUCCESS(rv, rv);
 
           nsIFrame *newCaretFrame = nullptr;
           nsCOMPtr<nsIContent> newCaretContent = startContent;
@@ -2279,8 +2286,8 @@
           do {
             // Continue getting the next frame until the primary content for the frame
             // we are on changes - we don't want to be stuck in the same place
-            frameTraversal.Next();
-            newCaretFrame = static_cast<nsIFrame*>(frameTraversal.CurrentItem());
+            frameTraversal->Next();
+            newCaretFrame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
             if (nullptr == newCaretFrame)
               break;
             newCaretContent = newCaretFrame->GetContent();
@@ -2696,14 +2703,21 @@
       continue;
     }
 
-    nsFrameIterator frameTraversal(presContext, startFrame,
-                                   ePreOrder, nsFrameIterator::FLAG_FOLLOW_OUT_OF_FLOW);
+    nsCOMPtr<nsIFrameEnumerator> frameTraversal;
+    nsresult rv = NS_NewFrameTraversal(getter_AddRefs(frameTraversal),
+                                       presContext, startFrame,
+                                       ePreOrder,
+                                       false, // aVisual
+                                       false, // aLockInScrollView
+                                       true      // aFollowOOFs
+                                       );
+    NS_ENSURE_SUCCESS(rv, rv);
 
     if (iterStartContent == aRootContent) {
       if (!aForward) {
-        frameTraversal.Last();
+        frameTraversal->Last();
       } else if (aRootContent->IsFocusable()) {
-        frameTraversal.Next();
+        frameTraversal->Next();
       }
     }
     else if (getNextFrame &&
@@ -2712,13 +2726,13 @@
       // Need to do special check in case we're in an imagemap which has multiple
       // content nodes per frame, so don't skip over the starting frame.
       if (aForward)
-        frameTraversal.Next();
+        frameTraversal->Next();
       else
-        frameTraversal.Prev();
+        frameTraversal->Prev();
     }
 
     // Walk frames to find something tabbable matching mCurrentTabIndex
-    nsIFrame* frame = static_cast<nsIFrame*>(frameTraversal.CurrentItem());
+    nsIFrame* frame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
     while (frame) {
       // TabIndex not set defaults to 0 for form elements, anchors and other
       // elements that are normally focusable. Tabindex defaults to -1
@@ -2788,10 +2802,10 @@
               Element* rootElement = subdoc->GetRootElement();
               nsIPresShell* subShell = subdoc->GetShell();
               if (rootElement && subShell) {
-                nsresult rv = GetNextTabbableContent(subShell, rootElement,
-                                                     aOriginalStartContent, rootElement,
-                                                     aForward, (aForward ? 1 : 0),
-                                                     false, aResultContent);
+                rv = GetNextTabbableContent(subShell, rootElement,
+                                            aOriginalStartContent, rootElement,
+                                            aForward, (aForward ? 1 : 0),
+                                            false, aResultContent);
                 NS_ENSURE_SUCCESS(rv, rv);
                 if (*aResultContent)
                   return NS_OK;
@@ -2835,10 +2849,10 @@
       // again.
       do {
         if (aForward)
-          frameTraversal.Next();
+          frameTraversal->Next();
         else
-          frameTraversal.Prev();
-        frame = static_cast<nsIFrame*>(frameTraversal.CurrentItem());
+          frameTraversal->Prev();
+        frame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
       } while (frame && frame->GetPrevContinuation());
     }
 